pkg/gather/gather.go

```diff
@@ -0,0 +1,254 @@
+			Errors:       result.Errs,
+			Panic:        result.Panic,
+		})
+	}
+
+	gathererMetadata := record.Record{
+		Name: fmt.Sprintf(gathererMetadataFilenamePattern, gathererName),
+		Item: record.JSONMarshaller{Object: GathererStatusReport{
+			StatusReports: functionReports,
+		}},
+	}
+	if err := rec.Record(gathererMetadata); err != nil {
+		errs = append(errs, fmt.Errorf("unable to record gatherer metadata because of error: %v", err))
+	}
+
+	var m runtime.MemStats
+	runtime.ReadMemStats(&m)
+
+	// if both gatherers get to the same archive we will get only the latest archive-metadata, which is totally fine
+	archiveMetadata := record.Record{
+		Name: archiveMetadataFilename,
+		Item: record.JSONMarshaller{Object: ArchiveMetadata{
+			MemoryAlloc:                m.HeapAlloc,
+			Uptime:                     time.Since(programStartTime).Truncate(time.Millisecond).Seconds(),
+			IsGlobalObfuscationEnabled: anonymizer != nil,
+		}},
+	}
+	if err := rec.Record(archiveMetadata); err != nil {
+		errs = append(errs, fmt.Errorf("unable to record archive metadata because of error: %v", err))
+	}
+
+	return sumErrors(errs)
+}
+
+// StartGatheringConcurrently starts gathering of enabled functions of the provided gatherer and returns a channel
+// with results which will be closed when processing is done
+func StartGatheringConcurrently(
+	ctx context.Context, gatherer Interface, enabledFunctions []string,
+) (chan GatheringFunctionResult, error) {
+	gathererName := gatherer.GetName()
+
+	gatherAllFunctions, gatherFunctionsList := GetListOfEnabledFunctionForGatherer(
+		gathererName, enabledFunctions,
+	)
+	if !gatherAllFunctions && len(gatherFunctionsList) == 0 {
+		return nil, fmt.Errorf("no gather functions are specified to run")
+	}
+
+	var tasks []Task
+
+	for _, functionName := range gatherer.ListGatheringFunctions() {
+		if !gatherAllFunctions && !utils.StringInSlice(functionName, gatherFunctionsList) {
+			continue
+		}
+
+		function, canFail, err := gatherer.GetGatheringFunction(functionName)
+		if err != nil {
+			klog.Errorf("Function %v not found in gatherer %v: %v", functionName, gathererName, err)
+			continue
+		}
+
+		tasks = append(tasks, Task{
+			Name:    functionName,
+			F:       function,
+			CanFail: canFail,
+		})
+	}
+
+	return HandleTasksConcurrently(ctx, tasks), nil
+}
+
+// GetListOfEnabledFunctionForGatherer parses a list of gathering functions to enable,
+// which has the following structure: []string {
+//   "clusterconfig/container_images",
+//   "clusterconfig/nodes",
+//   "clusterconfig/authentication",
+//   "othergatherer/some_function",
+// } where each item consists of a gatherer name and a function name split by a slash.
+// If there's a string "ALL", we enable everything and return the first parameter as true,
+// otherwise it will be false and the second parameter will contain function names
+func GetListOfEnabledFunctionForGatherer(gathererName string, allFunctionsList []string) (bool, []string) {
```

Comment:
Same here. Does this function have to be exported (starting with upper case letter)?

Reply:
Probably not